dosbox-staging: downgrade to 0.78.0

samford opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

`dosbox-staging` was updated to 0.78.1 (#92193) because upstream had published a `v0.78.1` tag (since removed). This used the stable tag format but there was no corresponding release on GitHub, no announcement on the homepage, etc.

Upstream confirmed this wasn't a release version and `v0.78.1-rc` would have been a more appropriate tag (dosbox-staging/dosbox-staging#1488). It's unclear when they will release a new version, so this PR downgrades the formula to the latest stable version, `0.78.0`. There aren't many installs (30 days: 90, 90 days: 183, 365 days: 537), so this shouldn't affect many users.

This PR also adds a `livecheck` block using the `GithubLatest` strategy, to align with the upstream release process (i.e., a version tag isn't a sufficient indicator of a new release).

Moisan reviewed (dismissed):
Good catch!

carlocab commented:
@chenrui333 can you check the release tags more carefully, please? This is the second time in the past month that we've had to look at downgrading because you approved a bad version bump. (cf. #91497, #92193)

CC @Homebrew/core

iMichka commented:
It's the hard part of `brew bump-formula`. When I approve these PRs, there are sometimes 10 to 20 open. And I do not check if the release is "official" for each of them.

I would mostly blame the issue on upstreams using stable tags when they should not: they can use RC or beta tags, but not stable tags.

So I think this issue will carry on, if it relies on manual verifying. Actually the whole `brew bump-formula` process is less usefull if we need to manually check if this is a real release or not.

iMichka commented:
Another issue we have is that we can't bump version schemes anymore, and this is a major pain.

MikeMcQuaid commented:
> Another issue we have is that we can't bump version schemes anymore, and this is a major pain.

@iMichka Can you elaborate on that?

iMichka commented:
@MikeMcQuaid I wanted to open an issue about this, but had no time. We already saw this in a previous PR where we tried to change the version scheme.
You will see the error here: https://github.com/Homebrew/homebrew-core/actions/runs/1655937944
It fails to upload the bottle because there seem to be an issue with the version scheme that is not taken into account.

MikeMcQuaid commented:
@iMichka That's not really what `version_scheme` was designed for so it's unsurprising. The version scheme hasn't changed here, we just want to go back to the previous version and force a downgrade. Doing that should not involve bumping the `version_scheme` or, if there's no other option now, you need to bump the `version_scheme` and `revision`.

carlocab commented:
I've always found using `version_scheme` to force downgrades a hack. I'd prefer avoiding having to bump `version_scheme` here, especially if there isn't really much of a difference between `0.78` and `0.78.1` (which it seems like there shouldn't be).

MikeMcQuaid commented:
> My thinking was that 0.78.1 was never actually released and the tag has since been deleted, so it doesn't feel appropriate for the formula to continue using a version that doesn't exist upstream if we don't know when the next release will be.

Sorry, be "force downgrade": I don't see why `brew upgrade` has to downgrade here. If a user has ended up on the wrong version: fine, they can `brew uninstall && brew install`.

carlocab commented:
> My thinking was that 0.78.1 was never actually released and the tag has since been deleted, so it doesn't feel appropriate for the formula to continue using a version that doesn't exist upstream if we don't know when the next release will be. It may be soon (tagging 0.78.1 could indicate it will be released in the near future) but it may also take a while (there's usually months between an unstable tag and the stable release).
> 
> 
> 
> Besides that, if we stay on "0.78.1", we won't be able to easily detect when a 0.78.1 stable version is released. The formula version and livecheck version would be equal, so it would look like the formula's up to date. It's fine if the deleted 0.78.1 tag and any later 0.78.1 tag reference the same commit but there's no guarantee of that (it looks like there are still open issues to resolve from 0.78.0).

Note that we can revert the original version bump without bumping `version_scheme`. I am in favour of reverting the version bump. We'll suffer some audit failures in the meantime, but that shouldn't be for too long since this formula is regularly updated.

samford commented:
Just so I understand, we simply want to remove the `version_scheme` bump from these changes, correct? If more changes are needed, let me know and I’ll take care of it when I’m back in front of my computer.

MikeMcQuaid commented:
> Ah, yeah, that makes sense to me. My intention here was simply to downgrade the formula, not necessarily to force users to switch versions on their next `brew upgrade`. Bumping `version_scheme` was the only way this was going to pass CI but if we're fine with the audit failures then I'm good to remove it.
> 
> Are any other changes needed here other than removing the `version_scheme` bump (e.g., reverting the bottle block)?

Yeh, I think remove the last change of the bottle block and `version_scheme`.

samford commented:
I removed `version_scheme`, reverted the `bottle` block, and rebased this for good measure. The audit failure is expected (i.e., `stable version should not decrease (from 0.78.1 to 0.78.0)`).
